### PR TITLE
Use TransitionInfo in more interfaces, part 2

### PR DESCRIPTION
**Pull request by wddgit**
#### PR description:

Use the TransitionInfo classes in more interfaces. In particular, it adds interface changes to streamTransitionAsync.h that were recently made in globalTransitionAsync.h so the two will remain similar. It also changes other interfaces from EventProcessor down the stack to the Worker classes and into the ProductResolvers.

The 6th commit percolates the transition info through the prefetchAsync functions in the ProductResolvers.

The 7th commit makes the UnscheduledAuxiliary store an EventTransitionInfo instead of the EventSetupImpl.

These two commits remove the need for const_casts that were in the UnscheduledProductResolver class and also remove the need to have a nullptr for the pointer to the EventSetupImpl in the ProcessBlock case.

(Note this is implemented on top of the changes in PR #30969. We could merge that one, then this one. Or just close that one and just handle it all in one PR. I could squash it all into one commit if you like. Let me know.)

@makortel @Dr15Jones You already reviewed the first 5 commits in the earlier PR and the subsequent commits deal with the unresolved comments. You had asked me to put these in a separate pull request as we were not sure how we wanted to deal with these. I had made a 3 part proposal in the old PR to resolve the issue. The first part went OK, but the other two parts did not work. The doWork function gets used in UnscheduledProductResolver in some corner cases. getProvenance can cause it to be called. So I could not delete that code. Instead I changed the UnscheduledAuxiliary to hold an EventTransitionInfo and that works. Another alternative would be to make sure in every case getProvenance that consumes is also called.

#### PR validation:

Ran existing unit tests. This should not change any behavior or any results. It just makes the code easier to read and understand.

### CODE REVIEW

**Comment by wddgit** [comment]
If I remove the call to doWork in UnscheduledProductResolver::resolveProduct_, then the following configuration fails. Other things may fail that also use getProvenance, but this is the test case I developed.

```
import FWCore.ParameterSet.Config as cms

process = cms.Process("TEST")

process.load("FWCore.Framework.test.cmsExceptionsFatal_cff")

process.load("FWCore.Modules.printContent_cfi")

process.printContent.listProvenance = cms.untracked.bool(True)

process.maxEvents = cms.untracked.PSet(
    input = cms.untracked.int32(5)
)

process.source = cms.Source("EmptySource")

process.Thing = cms.EDProducer("ThingProducer")

process.OtherThing = cms.EDProducer("OtherThingProducer")

process.ta = cms.Task(process.Thing)

process.p = cms.Path(process.printContent * process.OtherThing, process.ta)
```

**Comment by wddgit** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by wddgit** [comment]
@makortel @Dr15Jones I'm going to need some feedback to make more progress on this PR. The last two commits should be looked at separately to resolve the remaining two questions. At some level this is a question of style. There are multiple reasonable approaches that will work.

1. Do you like the percolation changes? One advantages are that some prefer the percolation style over the state variable in UnscheduledAuxiliary. Another advantage is if in the future we implement consumes calls for the cases where getProvenance is used, then we could delete UnscheduledAuxiliary entirely. The disadvantage is a small amount of extra complexity in the ProductResolver classes. To me either approach is reasonable.

The main alternative to percolation is using the TransitionInfo I added to UnscheduledAuxiliary in prefetchAsync_. Another alternative is using the const_cast. Another alternative is introducing a dummy EventSetupImpl. All of these will properly get the job done and the difference in performance is probably negligible.

2. Is the way I resolved the problem in UnscheduledProductResolver OK? The main thing is that the EventTransitionInfo is now stored in the UnscheduledAuxiliary. One alternative is to add consumes calls for the use cases of getProvenance. Then the doWork function could be removed entirely from UnscheduledProductResolver::resolveProduct_ and the UnscheduledAuxiliary class could be deleted. I like this alternative, but this is a big enough project it should probably be a separate PR (And there is some question whether it is worth the effort that will be involved. I don't understand how hard this will be.). Another alternative is to use the const_cast.

Separate from those questions. This PR could stand on its own and I could just close the earlier PR. Or do you want two separate PR's and I will rebase this one on top of the older one after it is merged. I also could squash out the multiple commits if you like (they add some changes back and forth that really add nothing useful to the history).

**Comment by wddgit** [comment]
Just wondering what to do with this. Do you like the percolation of the TransitionInfo in the prefetchAsync functions or should I back that out? Either way is OK with me. (See my previous comment, this percolation is in the 6th commit, the rest seems noncontroversial to me).

**Comment by makortel** [comment]
I have very mixed feelings and no clear opinion. 

> 1. Do you like the percolation changes? One advantages are that some prefer the percolation style over the state variable in UnscheduledAuxiliary. Another advantage is if in the future we implement consumes calls for the cases where getProvenance is used, then we could delete UnscheduledAuxiliary entirely. The disadvantage is a small amount of extra complexity in the ProductResolver classes.

I don't like much the changes in ProductResolvers caused by the percolation style, but on the other hand, I do think the percolation style is eventually clearer to reason than state variables. I have no idea at the moment if extending consumes to `getProvenance()` calls would be otherwise useful or not. @Dr15Jones can you comment?

> The main alternative to percolation is using the TransitionInfo I added to UnscheduledAuxiliary in prefetchAsync_. Another alternative is using the const_cast.

To clarify, in this PR the `EventTransitionInfo` is stored in `UnscheduledAuxiliary` similarly to `EventSetupImpl` in master, does that mean that this PR does not completely follow the percolation approach? I'd prefer to avoid `const_cast`.


> 2. Is the way I resolved the problem in UnscheduledProductResolver OK? The main thing is that the EventTransitionInfo is now stored in the UnscheduledAuxiliary.

It does not look much worse to me, at least. I'm a bit confused by the two calls to two `WorkerManager::setupOnDemandSystem()` overloads wrt. one call in code in master. 

> One alternative is to add consumes calls for the use cases of getProvenance. Then the doWork function could be removed entirely from UnscheduledProductResolver::resolveProduct_ and the UnscheduledAuxiliary class could be deleted. I like this alternative, but this is a big enough project it should probably be a separate PR (And there is some question whether it is worth the effort that will be involved. I don't understand how hard this will be.).

Removing `doWork()` call from `UnscheduledProductResolver:resolveProduct_()` sounds tempting, but I agree it would certainly be a subject of a separate PR, and I'm not sure if it would be worth it (perhaps not in the short term).

**Comment by wddgit** [comment]
In this PR, prefetchAsync_ uses percolation but resolveProduct_ does not.  If you look up the call stack you can see it is difficult to use percolation in the resolveProduct_ functions. I did not spend enough time studying the problem to understand it completely, but I think it would involve interface changes that affect code outside the Framework.

Both options are good. We shouldn't spend too much time thinking about it. Just choose and move on. I mildly prefer percolation, but am happy with both. Using the EventTransitionInfo stored in UnscheduledAuxiliary should also work well. It is simpler and if we never get around to doing the work needed to completely remove UnscheduledAuxiliary and the call to doWork, then it is the better choice.

**Comment by wddgit** [comment]
Matti asked: "It does not look much worse to me, at least. I'm a bit confused by the two calls to two WorkerManager::setupOnDemandSystem() overloads wrt. one call in code in master."

The existing function named setupOnDemandSystem internally has two parts which can be separated. By separating them, I managed to avoid making the function a template. One part is only used for events (not runs or lumis) and takes an EventSetupTransitionInfo. The other part is needed for runs, lumis and events but only needs a Principal (the base class, not RunPrincipal, LumiPrincipal, or EventPrincipal).

The name of this function has always bothered me because it has been been misnamed. The function does setup for several types of ProductResolvers, unscheduled, puttable, input, switch... In the case of puttable, input and switch resolvers, this has nothing to do with the "OnDemandSystem". Maybe I should rename the second part of this function "setupResolvers".

**Comment by makortel** [comment]
> In this PR, prefetchAsync_ uses percolation but resolveProduct_ does not.

Thanks, that clarifies.

> The existing function named setupOnDemandSystem internally has two parts which can be separated. By separating them, I managed to avoid making the function a template.

Makes sense (and I did think there was a good reason, it was just seeing two consecutive calls `setupOnDemandSystem()` with different arguments that looked a bit weird).

> Maybe I should rename the second part of this function "setupResolvers".

I agree, that would make the distinction of the two more clear.

File: FWCore/Framework/interface/Principal.h [hunk 1]
```diff
@@ -126,11 +129,17 @@ namespace edm {
...
-    void prefetchAsync(WaitingTask* waitTask,
+    template <typename INFOTYPE>
+    void prefetchAsync(WaitingTask* task,
                        ProductResolverIndex index,
                        bool skipCurrentProcess,
                        ServiceToken const& token,
-                       ModuleCallingContext const* mcc) const;
+                       ModuleCallingContext const* mcc,
+                       INFOTYPE const& info) const {
+      auto const& productResolver = productResolvers_.at(index);
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/Principal.h:139]: Please do not use `at` as this is performance critical code.

File: FWCore/Framework/src/ProductResolverBase.cc [hunk 2]
```diff
@@ -30,6 +30,18 @@ namespace edm {
...
 
+  void ProductResolverBase::connectTo(ProductResolverBase const&, EventTransitionInfo const&) { assert(false); }
+  void ProductResolverBase::connectTo(ProductResolverBase const&, LumiTransitionInfo const&) { assert(false); }
+  void ProductResolverBase::connectTo(ProductResolverBase const&, RunTransitionInfo const&) { assert(false); }
+  void ProductResolverBase::connectTo(ProductResolverBase const&, ProcessBlockTransitionInfo const&) { assert(false); }
+
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, EventTransitionInfo const&) const { assert(false); }
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, LumiTransitionInfo const&) const { assert(false); }
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, RunTransitionInfo const&) const { assert(false); }
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, ProcessBlockTransitionInfo const&) const {
```
Comment (Dr15Jones) [inline FWCore/Framework/src/ProductResolverBase.cc:41]: These defaults are telling me we are approaching this problem in the wrong way. A given ProductResolver will never be used by more than one kind of Principal. Therefore such a virtual overload set doesn't make any sense. It looks to me like the ProductResolver's should become templated on the Principal type for which they are going to be used.

Comment (makortel) [reply FWCore/Framework/src/ProductResolverBase.cc:41]: Should the `ProductResolverBase` be also templated on the `Principal` type?

Comment (Dr15Jones) [reply FWCore/Framework/src/ProductResolverBase.cc:41]: >Should the ProductResolverBase be also templated on the Principal type?

Yes, that is what I meant to say.

Comment (wddgit) [reply FWCore/Framework/src/ProductResolverBase.cc:41]: Does this also imply that both of you are saying you agree with the percolation approach? And I should go ahead and start working on making ProductResolverBase a template class. That seems a reasonable approach to me.

File: FWCore/Framework/interface/Principal.h (hunk 1 above)
Comment (makortel) [reply FWCore/Framework/interface/Principal.h:139]: It's already using `at()`, just in the `.cc`
https://github.com/cms-sw/cmssw/blob/53c50cdd24c436cca4892e8fb7e0d5b5b59b641e/FWCore/Framework/src/Principal.cc#L631

**Comment by wddgit** [comment]
Submitted another PR to replace this one.

**Pull request closed without merging**